pkg/dpm/manager.go

```diff
@@ -3,99 +3,217 @@ package dpm
-	defer dpm.fsWatcher.Close()
-	dpm.startPlugins()
+func (dpm *Manager) stopPluginServers(pluginMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
 
-	// The manager cannot be restarted, it is expected to be the `core` of an application.
-	<-dpm.stopCh
-	dpm.stopPlugins()
+	for pluginLastName, plugin := range pluginMap {
+		wg.Add(1)
+		go func() {
+			stopPluginServer(pluginLastName, plugin)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
 }
 
-// startPlugins is helper function to start the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) startPlugins() {
-	for _, plugin := range dpm.plugins {
-		go plugin.Start()
+func (dpm *Manager) stopPlugins(pluginMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
+	var pluginMapMutex = &sync.Mutex{}
+
+	for pluginLastName, plugin := range pluginMap {
+		wg.Add(1)
+		go func() {
+			stopPlugin(pluginLastName, plugin)
+			pluginMapMutex.Lock()
+			delete(pluginMap, pluginLastName)
+			pluginMapMutex.Unlock()
+			wg.Done()
+		}()
 	}
+	wg.Wait()
 }
 
-// startPlugins is helper function to stop the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) stopPlugins() {
-	for _, plugin := range dpm.plugins {
-		plugin.Stop()
+func startPlugin(pluginLastName string, plugin devicePlugin) {
+	if devicePluginImpl, ok := plugin.DevicePlugin.(PluginInterfaceStart); ok {
+		err := devicePluginImpl.Start()
+		if err != nil {
+			glog.Errorf("Failed to start plugin \"%s\": %s", pluginLastName, err)
+		}
 	}
+	startPluginServer(pluginLastName, plugin)
 }
 
-// handleSignals is the main signal handler that is responsible for handling of OS signals and Kubelet device plugins directory changes.
-func (dpm *DevicePluginManager) handleSignals() {
-	for {
-		select {
-		case s := <-dpm.signalCh:
-			switch s {
-			case syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGINT:
-				glog.V(3).Infof("Received signal \"%v\", shutting down", s)
-				close(dpm.stopCh)
-			}
-		case event := <-dpm.fsWatcher.Events:
-			if event.Name == pluginapi.KubeletSocket {
-				if event.Op&fsnotify.Create == fsnotify.Create {
-					dpm.startPlugins()
-				}
-				// TODO: Kubelet doesn't really clean-up it's socket, so this is currently manual-testing thing. Could we solve Kubelet deaths better?
-				if event.Op&fsnotify.Remove == fsnotify.Remove {
-					dpm.stopPlugins()
-				}
-			}
+func stopPlugin(pluginLastName string, plugin devicePlugin) {
+	stopPluginServer(pluginLastName, plugin)
+	if devicePluginImpl, ok := plugin.DevicePlugin.(PluginInterfaceStop); ok {
+		err := devicePluginImpl.Stop()
+		if err != nil {
+			glog.Errorf("Failed to stop plugin \"%s\": %s", pluginLastName, err)
+		}
+	}
+}
+
+func startPluginServer(pluginLastName string, plugin devicePlugin) {
+	for i := 1; i <= startPluginServerRetries; i++ {
+		err := plugin.StartServer()
+		if err == nil {
+			return
+		} else if i == startPluginServerRetries {
+			glog.V(3).Infof("Failed to start plugin server \"%s\", within given tries: %s", pluginLastName, err)
```

Comment:
Would be nice to log # of tries (since it'll have to be configurable).

Reply:
Done